eip/EIPS/eip-4337.md

```diff
@@ -252,23 +252,27 @@ There are 2 separate entry point methods: `handleOps` and `handleAggregatedOps`
 The entry point's `handleOps` function must perform the following steps (we first describe the simpler non-paymaster case). It must make two loops, the **verification loop** and the **execution loop**. In the verification loop, the `handleOps` call must perform the following steps for each `UserOperation`:
 
 * **Create the account if it does not yet exist**, using the initcode provided in the `UserOperation`. If the account does not exist, _and_ the initcode is empty, or does not deploy a contract at the "sender" address, the call must fail.
-* **Call `validateUserOp` on the account**, passing in the `UserOperation`, the required fee and aggregator (if there is one). The account should verify the operation's signature, and pay the fee if the account considers the operation valid. If any `validateUserOp` call fails, `handleOps` must skip execution of at least that operation, and may revert entirely.
+* calculate the maximum possible fee the account need to pay (based on validation and call gas limits, and current gas values)
+* calculate the fee the account must add to its "deposit" in the EntryPoint
+* **Call `validateUserOp` on the account**, passing in the `UserOperation`, its hash and the required fee. The account should verify the operation's signature, and pay the fee if the account considers the operation valid. If any `validateUserOp` call fails, `handleOps` must skip execution of at least that operation, and may revert entirely.
 * Validate the account's deposit in the entryPoint is high enough to cover the max possible cost (cover the already-done verification and max execution gas)
 
 In the execution loop, the `handleOps` call must perform the following steps for each `UserOperation`:
 
 * **Call the account with the `UserOperation`'s calldata**. It's up to the account to choose how to parse the calldata; an expected workflow is for the account to have an `execute` function that parses the remaining calldata as a series of one or more calls that the account should make.
+* After the call, refund the account's deposit with the excess gas price that was pre-charged.
+* After the execution of all calls, pay the collected fees from all UserOperations to the bundler's provided address
 
-![](../assets/eip-4337/image1.png)
+![](../assets/eip-4337/bundle-seq.svg)
 
-Before accepting a `UserOperation`, bundlers should use an RPC method to locally call the `simulateValidation` function of the entry point, to verify that the signature is correct and the operation actually pays fees; see the [Simulation section below](#simulation) for details.
+Before accepting a `UserOperation`, bundlers should use an RPC method to locally call the `simulateValidation` function on the entry point, to verify that the signature is correct and the operation actually pays fees; see the [Simulation section below](#simulation) for details.
```

Comment:
call the function *on* the entry point?

Reply:
it is a view function, in a separate contract...

Reply:
I know.  I just didn't understand how to apply the "on" part to a contract.  There's no clear way to "call f() on contract C" if `f()` is not defined in `C` so we're making a new terminology here :)

Reply:
We do need to add eip wording for the view functions. It's separate from the pr

Reply:
Will this be done as part of https://github.com/eth-infinitism/account-abstraction/pull/321 (the new simulation method)?

Reply:
yes. working on it there..